R/layer_templates.R

```diff
@@ -0,0 +1,321 @@
+#' `group_count(...)`. This is required, as after the template is built a user
+#' supplies these arguments via `use_template()`
+#'
+#' `use_template()` takes the `group_count|desc|shift()` arguments by default.
+#' If a user specified additional arguments in the template, these are provided
+#' in a list throught the argument `add_params`. Provide these arguments exactly
+#' as you would in a normal layer. When creating the template, these parameters
+#' can be specified by using curly brackets. See the examples for details.
+#'
+#' @param name Template name
+#' @param template Template layer syntax, starting with a layer constructor
+#'   `group_count|desc|shift`. This function should be called with an ellipsis
+#'   argument (i.e. group_count(...)).
+#'
+#' @md
+#' @export
+#'
+#' @family Layer Templates
+#' @rdname layer_templates
+#'
+#' @examples
+#'
+#' op <- options()
+#'
+#' new_layer_template(
+#'   "example_template",
+#'   group_count(...) %>%
+#'     set_format_strings(f_str('xx (xx%)', n, pct))
+#' )
+#'
+#' view_layer_templates()
+#'
+#' get_layer_template("example_template")
+#'
+#' tplyr_table(mtcars, vs) %>%
+#'   add_layer(
+#'     use_template("example_template", gear)
+#'   ) %>%
+#'   build()
+#'
+#' remove_layer_template("example_template")
+#'
+#' new_layer_template(
+#'   "example_template",
+#'   group_count(...) %>%
+#'     set_format_strings(f_str('xx (xx%)', n, pct)) %>%
+#'     set_order_count_method({sort_meth}) %>%
+#'     set_ordering_cols({sort_cols})
+#' )
+#'
+#' get_layer_template("example_template")
+#'
+#' tplyr_table(mtcars, vs) %>%
+#'   add_layer(
+#'     use_template("example_template", gear, add_params =
+#'                    list(
+#'                      sort_meth = "bycount",
+#'                      sort_cols = `1`
+#'                    ))
+#'   ) %>%
+#'   build()
+#'
+#' remove_layer_template("example_template")
+#'
+#' options(op)
+new_layer_template <- function(name, template) {
+  template <- enexpr(template)
+
+  # Have to convert the call to a character and collapse it to order text correctly
+  raw_template <- paste0(c(template), collapse="\n")
+
+  # Enforce ellipsis on layer constructors
+  if (!str_detect(raw_template, "group_(count|desc|shift)\\(\\.{3}\\)")) {
+    msg <- paste0(
+      "Invalid template - templates must start with an ellipsis (i.e. ...) passed to either ",
+      "group_count, group_desc, or group_shift. For example, group_count(...)"
+    )
+    stop(msg, call.=FALSE)
+  }
+
+  # Make sure that the template is valid
+  modify_nested_call(template, examine_only = TRUE)
+
+  if (name %in% names(getOption("tplyr.layer_templates"))) {
+    warning(
+      sprintf("A template by the name %s already exists. Template will be overwritten.", name),
+      call. = FALSE
+    )
+    remove_layer_template(name)
+  }
+
+  # Find any add_params provided into the template
+  params <- str_match_all(raw_template, "\\{\\s+([\\w\\.]+)\\s+\\}")[[1]][,2]
```

Comment:
The indexing here feels brittle to me. But I'm not sure if anything can be done since the result is an unnamed matrix.

Reply:
Comments added in latest push